exchange/exchange_test.go

```diff
@@ -2540,13 +2548,164 @@ func TestCategoryMappingTwoBiddersOneBidEachNoCategorySamePrice(t *testing.T) {
+
+	assert.Len(t, adapterBids[bidderNameApn1].bids, 0)
+	assert.Len(t, adapterBids[bidderNameApn2].bids, 2)
+
+	assert.Equal(t, "bid_idApn2_1", adapterBids[bidderNameApn2].bids[0].bid.ID, "Incorrect expected bid 1 id")
+	assert.Equal(t, "bid_idApn2_2", adapterBids[bidderNameApn2].bids[1].bid.ID, "Incorrect expected bid 2 id")
+
+	assert.Len(t, rejections, 2, "2 bids should be de-duplicated")
+	assert.Equal(t, "bid rejected [bid ID: bid_idApn1_1] reason: Bid was deduplicated", rejections[0], "Incorrect rejected bid 1")
+	assert.Equal(t, "bid rejected [bid ID: bid_idApn1_2] reason: Bid was deduplicated", rejections[1], "Incorrect rejected bid 2")
+
+}
+
+func TestRemoveBidByName(t *testing.T) {
+	cats1 := []string{"IAB1-3"}
+
+	bidApn1_1 := openrtb2.Bid{ID: "bid_idApn1_1", ImpID: "imp_idApn1_1", Price: 10.0000, Cat: cats1, W: 1, H: 1}
+	bidApn1_2 := openrtb2.Bid{ID: "bid_idApn1_2", ImpID: "imp_idApn1_2", Price: 20.0000, Cat: cats1, W: 1, H: 1}
+	bidApn1_3 := openrtb2.Bid{ID: "bid_idApn1_3", ImpID: "imp_idApn1_3", Price: 10.0000, Cat: cats1, W: 1, H: 1}
+	bidApn1_4 := openrtb2.Bid{ID: "bid_idApn1_4", ImpID: "imp_idApn1_4", Price: 20.0000, Cat: cats1, W: 1, H: 1}
+	bidApn1_5 := openrtb2.Bid{ID: "bid_idApn1_5", ImpID: "imp_idApn1_5", Price: 10.0000, Cat: cats1, W: 1, H: 1}
+	bidApn1_6 := openrtb2.Bid{ID: "bid_idApn1_6", ImpID: "imp_idApn1_6", Price: 20.0000, Cat: cats1, W: 1, H: 1}
+
+	bid1_Apn1_1 := pbsOrtbBid{&bidApn1_1, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}, nil, 0, false, ""}
+	bid1_Apn1_2 := pbsOrtbBid{&bidApn1_2, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}, nil, 0, false, ""}
+	bid1_Apn1_3 := pbsOrtbBid{&bidApn1_3, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}, nil, 0, false, ""}
+	bid1_Apn1_4 := pbsOrtbBid{&bidApn1_4, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}, nil, 0, false, ""}
+	bid1_Apn1_5 := pbsOrtbBid{&bidApn1_5, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}, nil, 0, false, ""}
+	bid1_Apn1_6 := pbsOrtbBid{&bidApn1_6, "video", nil, &openrtb_ext.ExtBidPrebidVideo{Duration: 30}, nil, 0, false, ""}
+
+	type aTest struct {
+		desc string
+		in   string
+	}
+	testCases := []aTest{
+		{
+			desc: "incorrect result after removing element from the middle",
+			in:   "bid_idApn1_3",
+		},
+		{
+			desc: "incorrect result after removing element from the end",
+			in:   "bid_idApn1_6",
+		},
+		{
+			desc: "incorrect result after removing element from the beginning",
+			in:   "bid_idApn1_1",
+		},
+	}
+	for _, test := range testCases {
+
+		innerBidsApn1 := []*pbsOrtbBid{
+			&bid1_Apn1_1,
+			&bid1_Apn1_2,
+			&bid1_Apn1_3,
+			&bid1_Apn1_4,
+			&bid1_Apn1_5,
+			&bid1_Apn1_6,
 		}
 
+		seatBidApn1 := &pbsOrtbSeatBid{bids: innerBidsApn1, currency: "USD"}
+
+		removeBidByName(seatBidApn1, test.in)
+		assert.Len(t, seatBidApn1.bids, 5, test.desc)
+		assert.False(t, checkBidPresentInArray(seatBidApn1.bids, test.in), "Bid should not be present in result")
```

Comment:
I think it's important to check we have the exact bids we expect. Rather than checking for the length and absence, I recommend testing for the complete collection. Something like:

```
{
	desc: "incorrect result after removing element from the middle",
	in:   "bid_idApn1_2",
	out:  []*pbsOrtbBid{&bid1_Apn1_1, &bid1_Apn1_3},
},
```

with the assertion
```
assert.ElementsMatch(t, test.out, seatBidApn1.bids, test.desc)
```

In which case the `checkBidPresentInArray` helper is no longer needed.

Reply:
Good idea! Modified!